### PR TITLE
Initial implementation of GCS artifact

### LINKED ISSUE #12: Investigate deploying plugins directly from the hub
https://docs.cdap.io/cdap/current/en/integrations/hub.html

Potentially a new resource "hub_artifact"

The main problem here is an artifact is multiple resources in the API (a JAR resource + properties). If one of the resources fails how do we recover from that and do we consider the entire resource broken?

Terraform has concept of "partial state". We could leverage that here.

### PR DESCRIPTION
**Pull request by umairidris**
Fix #12.

### CODE REVIEW

File: cdap/provider.go [hunk 1]
```diff
@@ -54,8 +57,9 @@ func Provider() *schema.Provider {
 
 // Config provides service configuration for service clients.
 type Config struct {
-	host   string
-	client *http.Client
+	host          string
+	client        *http.Client // TODO: rename to httpClient
```
Comment (xingao267) [inline cdap/provider.go]: why not rename to httpClient in this PR?

File: cdap/resource_gcs_artifact.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	"io/ioutil"
+	"regexp"
+
+	"cloud.google.com/go/storage"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+var bucketPathRE = regexp.MustCompile(`^gs://(.+)/(.+)$`)
+
+func resourceGCSArtifact() *schema.Resource {
```
Comment (xingao267) [inline cdap/resource_gcs_artifact.go]: Add a note why we need this and not directly use the TF's GCS object resource, and maybe in the future to migrate to use that one directly.

File: cdap/resource_gcs_artifact.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+		Value string `json:"value"`
+	} `json:"arguments"`
+}
+
+func resourceGCSArtifactCreate(d *schema.ResourceData, m interface{}) error {
+	ctx := context.Background()
+	config := m.(*Config)
+
+	matches := bucketPathRE.FindStringSubmatch(d.Get("bucket_path").(string))
+	if len(matches) != 3 {
```
Comment (xingao267) [inline cdap/resource_gcs_artifact.go]: why 3? I guess the spec, JAR and JSON. maybe add a doc above.

**Comment by xingao267** [comment]
Also add docs for this after https://github.com/GoogleCloudPlatform/terraform-provider-cdap/pull/17/files gets merged.

File: cdap/resource_gcs_artifact.go (hunk 2 above)
Comment (umairidris) [reply cdap/resource_gcs_artifact.go]: Done.

File: cdap/resource_gcs_artifact.go (hunk 3 above)
Comment (umairidris) [reply cdap/resource_gcs_artifact.go]: Done.

File: examples/artifact/main.tf [hunk 4]
```diff
@@ -27,31 +33,15 @@ provider "cdap" {
...
-  artifact_version = cdap_artifact.trash.version
-  value = "# Trash\n\nTrash consumes all the records on the input and eats them all,\nmeans no output is generated or no output is stored anywhere.\n"
-}
+# Option 2: Download JAR and JSON config and pass as local files.
+resource "cdap_local_artifact" "local_whistler_1_0_0" {
+  name = "whistler-transform"
+  version = "1.0.0"
+  json_config)path = "./TODO/whistler-transform-1.0.0.json"
+  jar_binary_path = "./TODO/whistler-transform-1.0.0.jar"
+}
```
[the remark this answers is not in the document]
Comment (umairidris) [reply examples/artifact/main.tf]: Done.

**Comment by umairidris** [comment]
I'll submit this first and then re gen the latest docs and submit the docs pr

**Pull request merged**